### PR TITLE
Implement Alibaba cloud SLS project

### PR DESCRIPTION
**Pull request by zzxwill**
### Description of your changes

Support provision Alibaba Cloud SLS project, create `examples` directory
for official examples and update makefile

Fixes #

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by negz** [comment]
> Hi, Nic, we're going to implement more cloud resources here

That's great news!

> do you have any suggestion about the priority of cloud resources needed?

Why don't we start an issue on this repository to gather input from users - Slack would be a great place to ask the community for input on this.

**Review by wonderflow — APPROVED** [review]

File: pkg/controller/sls/project_controller.go
Comment (negz) [inline pkg/controller/sls/project_controller.go:106]: @hasheddan Recently added some functionality to help extract credentials from a `ProviderConfig`. See https://github.com/crossplane/provider-template/blob/c38f3f9/internal/controller/mytype/mytype.go#L93 as an example.

File: pkg/controller/sls/project_controller.go
Comment (negz) [inline pkg/controller/sls/project_controller.go:117]: What is gained by breaking all this logic out into these separate `Base` functions, compared to just writing them inline?

**Review by negz — CHANGES_REQUESTED** [review]

**Comment by negz** [comment]
Thanks for working on this @zzxwill! Looks like a great start, but I would prefer if this managed resource was more in line with typical Crossplane conventions. If you strongly disagree with any we can discuss that and perhaps capture how provider-alibaba will diverge and why.

File: Makefile
Comment (zzxwill) [reply Makefile:148]: Thanks for this kind reminder. Done in #61

File: hack/boilerplate.go.txt
Comment (zzxwill) [inline hack/boilerplate.go.txt:2]: @negz I also moved this copyright matter to https://github.com/crossplane/provider-alibaba/pull/62. Thanks.

File: apis/sls/v1alpha1/project_types.go
Comment (zzxwill) [reply apis/sls/v1alpha1/project_types.go:38]: It's my default not to follow the conventions which is also mentioned in https://crossplane.io/docs/v0.5/services-developer-guide.html#finishing-touches. 
By the way, I moved these to `groupversion_info.go`, which is also supported by runtime. Thanks.

File: apis/sls/v1alpha1/project_types.go
Comment (zzxwill) [reply apis/sls/v1alpha1/project_types.go:57]: The docs is the API for Alibaba ROS (Resource Orchestration Service) and ROS will tag any service, like SLS or OSS. But for the SLS project management, only `Name` and `Description` are supported per https://www.alibabacloud.com/help/doc-detail/74780.htm, `Tag` is not supported by SLS API. Thanks.

File: pkg/clients/sls/project.go
Comment (zzxwill) [reply pkg/clients/sls/project.go:67]: Fixed.

File: pkg/clients/sls/project.go
Comment (zzxwill) [reply pkg/clients/sls/project.go:73]: Done.

File: pkg/controller/sls/base.go
Comment (zzxwill) [reply pkg/controller/sls/base.go:22]: Done, thanks.

File: pkg/controller/sls/base.go
Comment (zzxwill) [reply pkg/controller/sls/base.go]: So I need to upgrade to `v0.13.0`?  It doesn't work in version `v0.12.0`.

From the latest release note: 
> #242 Sets Deleting and Creating conditions in the managed reconciler to guard against a consumer forgetting to do so in the respective methods.

It might be better for me to do the upgrade in a separate PR as it will affect managed resource database and providerconfig.

File: apis/sls/v1alpha1/project_types.go
Comment (zzxwill) [reply apis/sls/v1alpha1/project_types.go]: Done, I've read the `high-fidelity` section and your article `Crossplane vs Terraform` in Chinese. They are great. thanks:)

File: pkg/controller/sls/base.go
Comment (zzxwill) [reply pkg/controller/sls/base.go:50]: Done. But I left `pkg/clients/sls/project.go` it be as I need to validate the type of the error like `NotExistxxx` issue and decided whether to redirect to create the external resource.

File: pkg/controller/sls/project_controller.go
Comment (zzxwill) [reply pkg/controller/sls/project_controller.go:106]: I will create another PR to utilized the common function. It needs much work:
- bump crossplane runtime version to `v0.13.0`
- throw around `ProviderConfig`
- refactor secret data for Alibaba provider
- use the common function to extract credentials and refactor `NewClient` parameters

File: pkg/controller/sls/project_controller.go
Comment (zzxwill) [reply pkg/controller/sls/project_controller.go:117]: Please kindly refer to issue #63, in which I tried to explain the reason in details and wrote a few examples.

File: pkg/controller/sls/project_controller.go
Comment (zzxwill) [reply pkg/controller/sls/project_controller.go:105]: Here is the PR https://github.com/crossplane/provider-alibaba/pull/64

File: pkg/controller/sls/base.go
Comment (negz) [reply pkg/controller/sls/base.go:75]: Ah, that's right. I see you have raised https://github.com/crossplane/provider-alibaba/pull/64 to address this. Thanks!

File: pkg/controller/sls/base.go
Comment (negz) [reply pkg/controller/sls/base.go:50]: Thanks - that sounds reasonable. Though FYI you can usually use `errors.Cause` in cases like this.

https://pkg.go.dev/github.com/pkg/errors?utm_source=godoc#Cause

File: pkg/controller/sls/base.go
Comment (negz) [inline pkg/controller/sls/base.go:73]: We should remove `cr.Spec.ForProvider.ProjectName` and instead use `meta.GetExternalName(cr)`. It turns out this pattern is very poorly documented, but we require it from all resources these days. You can see an example in the link below.

The `meta.GetExternalName()` should be used anywhere you currently use `cr.Spec.ForProvider.ProjectName`. function will return the value of the `crossplane.io/external-name` annotation, which Crossplane ensures is always set. If the user does not specify the annotation it is set automatically to the `metadata.name` of the `Project` managed resource. This allows Crossplane users to have consistent control over the "actual" or "external" name of their managed resources, rather than having to figure out which spec field corresponds to the name.

https://github.com/crossplane/provider-alibaba/blob/0cfdaf72d/pkg/controller/database/rdsinstance.go#L217

**Review by negz — COMMENTED** [review]
This is looking good @zzxwill, thank you. The main thing that still needs addressing is using `meta.GetExternalName`. I also wonder if we should agree on the pattern proposed in https://github.com/crossplane/provider-alibaba/issues/63 before we merge this PR, but if it's urgent I could be convinced to approve while we figure out https://github.com/crossplane/provider-alibaba/issues/63.

**Review by hongchaodeng — APPROVED** [review]
The code LGTM.

Please address @negz 's comments. @negz Please help approve when it is done :)

**Comment by wonderflow** [comment]
HI @negz  Can we merge it first and fix the rest in following PRs. We need it to be merged and show in demos.

**Comment by negz** [comment]
> HI @negz Can we merge it first and fix the rest in following PRs. We need it to be merged and show in demos.

No, sorry. I recommend building the provider from a fork if you need to use it for demos. I'm happy to merge once the remaining comments are addressed.

File: pkg/controller/sls/base.go
Comment (zzxwill) [reply pkg/controller/sls/base.go:51]: `errors.Cause` is more reasonable which can make all errors in sls project client more readable. So I fixed all of them.

**Comment by zzxwill** [comment]
> This is looking good @zzxwill, thank you. The main thing that still needs addressing is using `meta.GetExternalName`. 
Done. Per your suggestion, I believe SLS project name `spec.froProvider.name` should be deleted and the managed resource name should be regarded as the SLS project name.

> I also wonder if we should agree on the pattern proposed in #63 before we merge this PR, but if it's urgent I could be convinced to approve while we figure out #63.
I will add more some examples in #63 after I talked with my colleagues which will integrated provider-base base logics.

**Comment by zzxwill** [comment]
@negz And I am so sorry for the late of addressing your newly two comments:) I was so busy with some other stuffs.

**Comment by negz** [comment]
No problem @zzxwill! We've been very busy too. :) Thanks for addressing my comments about using `GetExternalName` and `pkg/errors`. It seems like there are now only two things we need to sort out to merge this:

* The latest commit is failing our linter; I think one of the files needs a `gofmt`.
* Figuring out what we are going to do about https://github.com/crossplane/provider-alibaba/issues/63.
 
I prefer to not introduce the `BaseObserve` pattern in this PR until _after_ we have a chance to discuss the context and approach in https://github.com/crossplane/provider-alibaba/issues/63. How do you feel about updating this PR to use a 'normal' `ExternalClient` implementation for now, until we discuss further and agree?

**Comment by zzxwill** [comment]
@negz Your suggestion is helpful for decomposing the problems. Addressed all of the two. Thanks help review it:)

**Comment by zzxwill** [comment]
> Why don't we start an issue on this repository to gather input from users - Slack would be a great place to ask the community for input on this.

Create the issue #67, and I will talk to some potential users and contributors offline first.

**Review by negz — APPROVED** [review]
Excellent work @zzxwill - thank you!

**Pull request merged**
